app/lib/logstash_logging.rb

```diff
@@ -1,49 +1,80 @@
+require 'request_store_rails'
+
 class LogstashLogging
   DOMAIN_FOR_LOGS = Rails.env.production? ? AzureEnvironment.hostname : Socket.gethostname
+  SERVICE_NAME = ENV['SERVICE_NAME'] # e.g. web, worker or clock
 
-  def self.enable(config)
-    config.log_level = :info # :debug does not make sense with lograge + logstash
-
-    # Use default logging formatter so that PID and timestamp are not suppressed.
-    config.log_formatter = ::Logger::Formatter.new
+  def self.add_id_fields(event)
+    candidate_id = RequestLocals.fetch(:candidate_id) { nil }
+    if candidate_id
+      event['candidate_id'] = candidate_id
+    end
+    vendor_api_token_id = RequestLocals.fetch(:vendor_api_token_id) { nil }
+    # TODO: is there a vendor_id we should also log?
+    if vendor_api_token_id
+      event['vendor_api_token_id'] = vendor_api_token_id
+      event['provider_id'] = RequestLocals.fetch(:provider_id) { nil }
+    end
+  end
 
-    # Prepend all STDOUT log lines with the following tags
-    config.log_tags = [:request_id]
+  def self.add_sidekiq_fields(event)
+    tid = Thread.current['sidekiq_tid']
+    if !tid.blank?
+      ctx = Sidekiq::Context.current
+      event['tid'] = tid
+      event['ctx'] = ctx
+    end
+  end
 
-    # Make stdout_logger the standard Rails logger
-    stdout_logger = ActiveSupport::Logger.new(STDOUT)
-    stdout_logger.formatter = config.log_formatter
-    config.logger = ActiveSupport::TaggedLogging.new(stdout_logger)
+  def self.enable(rails_config)
+    # Add domain, params etc. to the logs
```

Comment:
This is gnarly! (the comment is an understatement 😄)

I feel comfortable with the simple constant-to-hash assignments, but the attribution stuff feels like too much to do in this file.

There are a few cross-cutting concerns in this method: 

- mapping `domain`, `service` and `params`
- mapping the user identity (`candidate`, `provider_id` etc)
- mapping the `exception` field (in a different way to the above)
- configuring lograge

In the case of the identity mapping, is there a case for this hash being provided by the controller in the way that `add_candidate_id_to_log` does? Could we simplify the logstash configuration by forcing controllers to implement some interface like `add_user_identity_to_log`, have them set something like

```
RequestLocals.store[:identity] = {candidate_id: foo}
```

and then in the logstash merge that in: `event.merge(RequstLocals.store[:identity])`.

Reply:
Please previous comment about logstash-logger not having access to a controller context. Lograge does, and in previous iterations I had such methods in the relevant base controllers. With logstash-logger, however, the only communication strategy between the controller and the logger is through ```request_store_rails``` or equivalent.

I have kept lograge because it gives us the one line per request, does away with ActionView logging (which partials are being rendered etc.) and has access to 'exception' in case of errors.

But I've moved all other attribution to logstash-logger, because I couldn't get lograge to affect sidekiq and clockwork logs.

I like your 'identity' event.merge suggestion, will update.